### PR TITLE
Add configurable retries

### PR DESCRIPTION
**Pull request by zhumuht**
add retry option.  when retry_times less than max_error_times, not set available = False

### CODE REVIEW

File: custom_components/xiaomi_miio_fan/fan.py [hunk 1]
```diff
@@ -24,9 +24,11 @@
 _LOGGER = logging.getLogger(__name__)
 
 DEFAULT_NAME = 'Xiaomi Miio Fan'
+DEFAULT_MAX_ERROR_TIMES = 20
```
Comment (syssi) [inline custom_components/xiaomi_miio_fan/fan.py]: Please call it `DEFAULT_RETRIES`.

File: custom_components/xiaomi_miio_fan/fan.py [hunk 2]
```diff
@@ -24,9 +24,11 @@
...
 DATA_KEY = 'fan.xiaomi_miio_fan'
 
 CONF_MODEL = 'model'
+CONF_RETRY = 'retry'
```
Comment (syssi) [inline custom_components/xiaomi_miio_fan/fan.py]: `CONF_RETRIES = 'retries'`

File: custom_components/xiaomi_miio_fan/fan.py [hunk 3]
```diff
@@ -186,6 +189,7 @@
     name = config.get(CONF_NAME)
     token = config.get(CONF_TOKEN)
     model = config.get(CONF_MODEL)
+    max_error_times = config.get(CONF_RETRY)
```
Comment (syssi) [inline custom_components/xiaomi_miio_fan/fan.py]: `retries = config.get(CONF_RETRIES)`

File: custom_components/xiaomi_miio_fan/fan.py [hunk 4]
```diff
@@ -253,12 +257,14 @@
...
     """Representation of a generic Xiaomi device."""
 
-    def __init__(self, name, device, model, unique_id):
+    def __init__(self, name, device, model, unique_id, max_error_times):
         """Initialize the generic Xiaomi device."""
         self._name = name
         self._device = device
         self._model = model
         self._unique_id = unique_id
+        self._error_times = 0
```
Comment (syssi) [inline custom_components/xiaomi_miio_fan/fan.py]: `self._retry = 0`

File: custom_components/xiaomi_miio_fan/fan.py [hunk 5]
```diff
@@ -253,12 +257,14 @@
...
+        self._max_error_times = max_error_times
```
Comment (syssi) [inline custom_components/xiaomi_miio_fan/fan.py]: `self._retries = retries`

**Comment by zhumuht** [comment]
> Could you rename some variables? Thanks! I will merge your changes afterwards.

done. thanks for your suggestion:)

File: custom_components/xiaomi_miio_fan/fan.py [hunk 6]
```diff
@@ -442,10 +447,15 @@ def supported_features(self) -> int:
...
 
         except DeviceException as ex:
-            self._available = False
-            _LOGGER.error("Got exception while fetching the state: %s", ex)
+            self._retry = self._retry + 1
+            if self._retry < self._retries:
+                _LOGGER.warning("Got exception while fetching the state: %s , _retry=%s", ex, self._retry)
+            else:
+                self._available = False
+                _LOGGER.error("Got exception while fetching the state: %s , _retry=%s", ex, self._retry)
```
Comment (syssi) [inline custom_components/xiaomi_miio_fan/fan.py]: The log message is equal in both cases. Please invert the if condition and move the log message above.

Comment (syssi) [reply custom_components/xiaomi_miio_fan/fan.py]: Ping! :-)

**Pull request merged**